python/lsst/ts/ess/csc/data_client/siglent_ssa3000x_spectrum_analyzer_data_client.py

```diff
@@ -0,0 +1,221 @@
+
+    @property
+    def connected(self) -> bool:
+        return self.client is not None and self.client.connected
+
+    async def connect(self) -> None:
+        await self.disconnect()
+
+        if self.simulation_mode == 0:
+            self.client = tcpip.Client(host=self.config.host, port=self.config.port)
+            await asyncio.wait_for(self.client.start_task, self.config.connect_timeout)
+        else:
+            self.log.info(
+                "Simulating output from an SSA3000X Spectrum Analyzer serial interface at "
+                f"host={self.config.host}, port={self.config.port}"
+            )
+
+    async def disconnect(self) -> None:
+        self.read_loop_task.cancel()
+        try:
+            if self.connected:
+                assert self.client is not None  # make mypy happy
+                await self.client.close()
+        finally:
+            self.client = None
+
+    async def run(self) -> None:
+        self.read_loop_task.cancel()
+        self.read_loop_task = asyncio.create_task(self.read_loop())
+
+    async def write(self, data: str) -> None:
+        """Write the data appended with a newline character.
+
+        Parameters
+        ----------
+        data : `str`
+            The data to write.
+        """
+        assert self.client is not None  # make mypy happy
+        await self.client.write(data.encode() + b"\r\n")
+
+    async def read_loop(self) -> None:
+        """Read raw data from the SSA3000X Spectrum Analyzer."""
+        starttime = datetime.now()
+        try:
+            if self.connected and self.simulation_mode == 0:
+                # Make sure that the correct frequency range is used by the
+                # spectrum analyzer.
+                await self.write(self.set_freq_start_cmd)
+                await self.write(self.set_freq_stop_cmd)
+            while self.connected or self.simulation_mode != 0:
+                timestamp = current_tai()
+                if self.simulation_mode == 0:
+                    await self.write(self.query_trace_data_cmd)
+                    assert self.client is not None  # make mypy happy
+                    read_bytes = await asyncio.wait_for(
+                        self.client.read_until(TERMINATOR),
+                        timeout=self.config.read_timeout,
+                    )
+                    raw_data = read_bytes.decode().strip()
+                    raw_data_items = raw_data.split(",")
+                    data = [float(i.strip()) for i in raw_data_items]
+                else:
+                    # Generate random data between 0 and -100 dB. Convert to
+                    # list to keep mypy happy.
+                    data = (-100.0 * np.random.random(self.num_data_points)).tolist()
+                try:
+                    await self.topics.tel_spectrumAnalyzer.set_write(
+                        frequency=self.frequencies, trace=data, timestamp=timestamp
+                    )
```

Comment:
Does this timestamp field add any useful information? I see that most of our topics have it, but I'm not convinced they ought to (see my comment on ts_xml). In any case, does it add information here? It seems to me that it won't be any more accurate than private_sndStamp, unless the spectrum analyzer is very slow to return data *and* the start time is better in some way then the end time.